Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,601 @@
+        /// `ClosedConnectionRemoveAction` which instructs it about what it should do.
+        private func configureCloseCallback(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { result in
+                switch result {
+                case .success:
+                    let action = self.parentPool.connectionProvidersLock.withLock {
+                        self.stateLock.withLock { self.state.removeClosedConnection(connection) }
+                    }
+                    switch action {
+                    case .none:
+                        break
+                    case .makeConnectionAndComplete(let el, let promise):
+                        self.makeConnection(on: el).cascade(to: promise)
+                    }
+
+                case .failure(let error):
+                    preconditionFailure("Connection close future failed with error: \(error)")
+                }
+            }
+        }
+
+        func syncClose(requiresCleanClose: Bool) throws {
+            var closeError: Error?
+            let availableConnections = self.stateLock.withLock { () -> CircularBuffer<ConnectionPool.Connection> in
+                assert(!self.state.isClosed, "Calling syncClose on an already closed provider")
+                let waitersCopy = self.state.waiters
+                self.state.waiters.removeAll()
+                for waiter in waitersCopy {
+                    waiter.promise.fail(HTTPClientError.cancelled)
+                }
+                self.state.isClosed = true
+                if requiresCleanClose {
+                    // TODO: This should probably become an assertion
+                    if self.state.leased != 0 {
+                        closeError = HTTPClientError.uncleanShutdown
+                    }
+                }
+                return self.state.availableConnections
+            }
+            do {
+                try EventLoopFuture<Void>.andAllComplete(availableConnections.map { $0.channel.close() }, on: self.eventLoop).wait()
+            } catch ChannelError.alreadyClosed {
+                return
+            } catch {
+                // First catched errors have priority, this
+                // is to be consistent with what would happen
+                // if the `throw` wasn't delayed until everything
+                // is executed
+                if closeError == nil {
+                    closeError = error
+                }
+            }
+
+            self.stateLock.withLock { self.state.availableConnections.removeAll() }
+
+            if let closeError = closeError {
+                throw closeError
+            }
+        }
+
+        // FIXME: Should this be a precondition or an assertion?
+        private func preconditionIsOpened() {
+            self.stateLock.withLock {
+                precondition(self.state.isClosed == false, "Attempting to use closed HTTP1ConnectionProvider")
+            }
+        }
```

Comment:
@weissi @Lukasa I wonder if this should be a `precondition` or an `assertion`? Is this error worth crashing the executable even in production?

Reply:
@adtrevor I think `precondition` is good here

Reply:
if your internal state is really broken, then a crash is better than just trying to continue. We'll get bug reports which we can then address

Reply:
Thanks